ext/mysqlnd/mysqlnd_auth.c

```diff
@@ -490,24 +487,15 @@ mysqlnd_auth_change_user(MYSQLND_CONN_DATA * const conn,
 		}
 	}
 	if (ret == PASS) {
-		char * tmp = NULL;
-		/* if we get conn->username as parameter and then we first free it, then estrndup it, we will crash */
-		tmp = mnd_pestrndup(user, user_len, conn->persistent);
-		if (conn->username.s) {
-			mnd_pefree(conn->username.s, conn->persistent);
-		}
-		conn->username.s = tmp;
+		mysqlnd_set_persistent_string(&conn->username, user, user_len, conn->persistent);
 
-		tmp = mnd_pestrdup(passwd, conn->persistent);
+		char * tmp = mnd_pestrdup(passwd, conn->persistent);
```

Comment:
Shouldn't the conn->password assignment be changed as well?

Reply:
It's the non-N version. I wasn't sure if I can touch this given that it is a variable-length assignment and the previous code didn't set the `.l` property anywhere. Is it safe to use `strlen()` here?

Reply:
pestrdup = pestrndup + strlen basically. You make a good point on this not assigning `.l` though (also applies to username assignment above), which seems suspect to me. I assume the length just never gets used?

Reply:
I noticed there is `passwd_len` in the signature, so I used it. Is this ok?

I grepped for `conn->password` but I can't see any usage of this property. To be frank, I don't understand what is the purpose of these properties on the `conn` object. The connection seems to happen before they are saved. The `change_user` method also uses the values from the signature and then copies the strings.

Reply:
Good point. Let's land this patch first and then remove them. I could see how they might be useful for a hypothetical "automatic reconnect" functionality, but given how such a functionality doesn't exist...